beets/autotag/mb.py

```diff
@@ -50,7 +56,256 @@
     "year": "date",
 }
 
-musicbrainzngs.set_useragent("beets", beets.__version__, "https://beets.io/")
+# Rate limiting
+
+limit_interval = 1.0
+limit_requests = 1
+do_rate_limit = True
+
+
+def set_rate_limit(limit_or_interval=1.0, new_requests=1):
+    """Sets the rate limiting behavior of the module. Must be invoked
+    before the first Web service call.
+    If the `limit_or_interval` parameter is set to False then
+    rate limiting will be disabled. If it is a number then only
+    a set number of requests (`new_requests`) will be made per
+    given interval (`limit_or_interval`).
+    """
+    global limit_interval
+    global limit_requests
+    global do_rate_limit
+    if isinstance(limit_or_interval, bool):
+        do_rate_limit = limit_or_interval
+    else:
+        if limit_or_interval <= 0.0:
+            raise ValueError("limit_or_interval can't be less than 0")
+        if new_requests <= 0:
+            raise ValueError("new_requests can't be less than 0")
+        do_rate_limit = True
+        limit_interval = limit_or_interval
+        limit_requests = new_requests
```

Comment:
I can undertsand that ! So far the PR s a 1st step. That's a simple copy from what used to be in musicbrainzngs.

If the whole idea is fine for beets, I'm ok to find a solution. Would you have any tip for the singleton ?

Reply:
I would look up singletons in Python, but they're pretty simple. If you don't know what they are (and sorry for being presumptive, if you do know then ignore), they are classes that can only create a single instance. You store an instance in the class data and that is the only one that is every created, which allows you to keep track of state like with these global variables, but in an OOP way.

[Here](https://www.geeksforgeeks.org/singleton-pattern-in-python-a-complete-guide/) is an explanation of how to create a singleton in Python. Again, ignore if you're already aware.

For the implementation, I would suggest perhaps a class that manages requests and handles the rate limiting. The RateLim decorator would be the best place, you can make it a singleton and then just use the decorator. If that doesn't work, then there can be a separate object that manages the rate inside the decorator should be fine. It's basically just trying to keep track of a counter and maybe a lock for the sake of concurrency. That's prime singleton stuff.

Reply:
Also, as far as first steps go, this is good! My idea on PRs is that they should be complete enough to not need immediate changes. I think the global variables would be that type of thing; it's just a change that should be done, and there's no reason not to do it now rather than add them and then have to remove them in a separate PR.